File: packages/core/src/transport/ulid.ts
// Monotonic ULID generation for client→server envelope `id`.
//
// Hand-rolled because `@dhaam-ccrm/core` has zero runtime dependencies —
// no `ulid` package, no `uuid`, no crypto polyfill.
//
// Why monotonic specifically (rather than plain random ULIDs): a ULID's
// sort order is its timestamp prefix followed by its random suffix. Two
// ULIDs minted in the SAME millisecond have identical prefixes, so their
// relative order is decided by 80 bits of unrelated randomness — it flips
// run to run. Sending three messages in one tick would give three ids
// whose lexical order does not match the order the user typed them. The
// ULID spec's monotonic factory fixes this by reusing the previous random
// suffix incremented by one whenever the clock has not advanced:
//
//   "When generating a ULID within the same millisecond, we can provide
//    some guarantees regarding sort order. Namely, if the same millisecond
//    is detected, the `random` component is incremented by 1 bit in the
//    least significant bit position (with carrying)."
//   — https://github.com/ulid/spec#monotonicity
//
// This matters beyond aesthetics: for `message.send` the envelope `id` IS
// the message's permanent id (D1, PRD §0.5/§9.3), so an unstable order
// here is an unstable order in persisted data.
//
// The server enforces the same 26-char Crockford base32 shape on EVERY
// client frame's `id`, not just `message.send` — see protocol/validate.ts
// `isValidUlid`, which is byte-identical to the server's own pattern.

import type { Clock } from '../presence/time.js';

/**
 * Returns a float in `[0, 1)`, exactly like `Math.random`.
 *
 * Injectable so tests can pin the random suffix and assert monotonic
 * increment without relying on chance.
 */
export type RandomSource = () => number;

/** Mints one ULID per call. */
export type UlidGenerator = () => string;

/** Crockford base32 — the standard alphabet minus I, L, O and U. */
const ENCODING = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
const ENCODING_LEN = 32;

/** 10 base32 chars of 48-bit millisecond timestamp. */
const TIME_CHARS = 10;

/** 16 base32 chars of 80-bit randomness. */
const RANDOM_CHARS = 16;

/** 2^48 - 1 — the largest timestamp representable in 10 base32 chars. */
const MAX_TIME_MS = 281_474_976_710_655;

/** Largest value a single base32 char can hold. */
const MAX_CHAR_VALUE = ENCODING_LEN - 1;

/**
 * Reads `globalThis.crypto` lazily, per call — never at module scope.
 * Importing this module must not touch any global (PRD §14, and the same
 * SSR-safety rule that governs the `WebSocket` global in socket.ts).
 * Falls back to `Math.random` where WebCrypto is absent.
 */
const defaultRandom: RandomSource = () => {
  const webCrypto = (
    globalThis as {
      crypto?: { getRandomValues?: (array: Uint32Array) => Uint32Array };
    }
  ).crypto;

  if (typeof webCrypto?.getRandomValues === 'function') {
    const buffer = new Uint32Array(1);
    webCrypto.getRandomValues(buffer);
    return (buffer[0] ?? 0) / 0x1_0000_0000;
  }

  return Math.random();
};

/** Encodes epoch millis as `TIME_CHARS` base32 chars, most significant first. */
function encodeTime(timeMs: number): string {
  let remaining = timeMs;
  let out = '';

  for (let i = 0; i < TIME_CHARS; i += 1) {
    const digit = remaining % ENCODING_LEN;
    // `charAt` rather than `[]` — under `noUncheckedIndexedAccess` an index
    // access is `string | undefined`, and `digit` is provably in range.
    out = ENCODING.charAt(digit) + out;
    remaining = (remaining - digit) / ENCODING_LEN;
  }

  return out;
}

/** Draws a fresh 80-bit random suffix as `RANDOM_CHARS` 5-bit digits. */
function drawRandom(random: RandomSource): number[] {
  const digits: number[] = [];

  for (let i = 0; i < RANDOM_CHARS; i += 1) {
    // `Math.floor` of a `[0, 1)` float scaled by 32 lands in `[0, 31]`.
    // `Math.min` guards a `RandomSource` that returns exactly 1.
    digits.push(Math.min(Math.floor(random() * ENCODING_LEN), MAX_CHAR_VALUE));
  }

  return digits;
}

/**
 * Adds one to the 80-bit suffix in place, with carry.
 *
 * Returns `false` on overflow — every one of the 16 digits was already at
 * its maximum, i.e. 2^80 ULIDs were minted inside a single millisecond.
 */
function incrementRandom(digits: number[]): boolean {
  for (let i = digits.length - 1; i >= 0; i -= 1) {
    const digit = digits[i] ?? 0;

    if (digit < MAX_CHAR_VALUE) {
      digits[i] = digit + 1;
      return true;
    }

    digits[i] = 0;
  }

  return false;
}

function encodeRandom(digits: number[]): string {
  let out = '';
  for (const digit of digits) out += ENCODING.charAt(digit);
  return out;
}

export interface UlidGeneratorOptions {
  /** Defaults to `Date.now`. */
  now?: Clock;

  /** Defaults to WebCrypto when present, `Math.random` otherwise. */
  random?: RandomSource;
}

/**
 * Builds a monotonic ULID generator.
 *
 * Guarantees, in order of importance:
 *
 * 1. Every returned string matches protocol/validate.ts's `isValidUlid`
 *    (26 chars, Crockford base32, uppercase) — the same check the server
 *    applies to every inbound client frame.
 * 2. Successive calls are strictly increasing lexically, including within
 *    one millisecond and across a clock that jumps *backwards* (NTP
 *    correction, VM resume). A backwards jump reuses the last timestamp
 *    rather than emitting an id that sorts before its predecessor.
 *
 * @throws RangeError if the clock reads negative, non-finite, or beyond
 *   2^48-1 ms — an id built from such a clock could not round-trip.
 * @throws RangeError on 80-bit suffix overflow inside one millisecond
 *   (unreachable in practice; ~1.2e24 ids per ms would be required). The
 *   ULID spec mandates an error here rather than silently wrapping, which
 *   would break guarantee 2.
 */
export function createUlidGenerator(options: UlidGeneratorOptions = {}): UlidGenerator {
  const { now, random = defaultRandom } = options;
  const clock: Clock = now ?? ((): number => Date.now());

  let lastTimeMs = -1;
  let lastRandom: number[] = [];

  return function nextUlid(): string {
    const rawNow = clock();

    if (!Number.isFinite(rawNow) || rawNow < 0 || rawNow > MAX_TIME_MS) {
      throw new RangeError(`ULID timestamp out of range: ${String(rawNow)}`);
    }

    const timeMs = Math.floor(rawNow);

    // `<=`, not `===`: a backwards clock must not be allowed to emit an id
    // that sorts before the previous one. Pinning to `lastTimeMs` and
    // incrementing keeps the sequence strictly increasing either way.
    if (timeMs <= lastTimeMs) {
      if (!incrementRandom(lastRandom)) {
        throw new RangeError('ULID random component overflowed within a single millisecond');
      }
    } else {
      lastTimeMs = timeMs;
      lastRandom = drawRandom(random);
    }

    return encodeTime(lastTimeMs) + encodeRandom(lastRandom);
  };
}

File: packages/dart/lib/src/protocol/ulid.dart
/// ULID generation and validation — PRD §0.5 D1, §9.3.
///
/// D1 makes the client-generated ULID the message's PERMANENT id: the server
/// validates the format, enforces per-session uniqueness, and stores it as the
/// canonical id. There is no id-swap path anywhere in this package. That makes
/// this file load-bearing in a way it would not be if the server minted ids —
/// a malformed id here is a message that can never be sent, and a colliding
/// one is a message that silently vanishes into a dedup check.
///
/// ── SPEC GAP (recorded deliberately) ──────────────────────────────────────
///
/// §7.2 says the envelope `id` is a "ULID" and D1 says the server "validates
/// ULID format". Neither §7 nor the OpenAPI document states WHAT that format
/// is: not the length, not the alphabet, not the case, not the time/random
/// split, and not whether monotonicity within a millisecond is required.
///
/// The accepted format is `^[0-9A-HJKMNP-TV-Z]{26}$` — 26 characters of
/// uppercase Crockford base32 (the alphabet omits I, L, O and U to survive
/// transcription). That regex is enforced by the server and by the TypeScript
/// core, and appears in neither the PRD nor the OpenAPI spec. An implementer
/// working from §7 alone would plausibly reach for a UUID, or a lowercase
/// ULID, and every frame they sent would be refused with VALIDATION_FAILED.
///
/// Deriving this required reading the implementation. See README.md §"Spec
/// gaps found".
library;

import 'dart:math';

/// Crockford base32, excluding I, L, O and U.
const String _encoding = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

/// 10 characters of timestamp + 16 of randomness = the canonical 26.
const int _timeLength = 10;
const int _randomLength = 16;

/// The one canonical ULID shape. Must stay byte-identical to the server's
/// `ULID_PATTERN`; loosening it here does not loosen it there.
final RegExp _ulidPattern = RegExp(r'^[0-9A-HJKMNP-TV-Z]{26}$');

/// Whether [value] is a syntactically valid ULID.
///
/// This is the same test the server applies to every inbound envelope `id`,
/// so anything failing here would be refused on the wire.
bool isValidUlid(String value) => _ulidPattern.hasMatch(value);

/// Generates ULIDs that sort lexicographically in creation order.
///
/// ── Why this is a class and not a top-level function ──────────────────────
///
/// Monotonicity within a millisecond requires remembering the previous value.
/// A ULID's first 10 characters encode epoch-millis, so two ULIDs minted in
/// the same millisecond differ only in their random half and sort in random
/// order relative to each other. Messages are ordered by `seq` on the wire
/// (D2), so this does not affect protocol correctness — but a host app that
/// sorts its own optimistic, not-yet-acked messages by id would see them
/// flicker into a different order, and two messages sent by a fast typist land
/// in the same millisecond routinely.
///
/// So when the clock has not advanced, the random half is incremented as a
/// base-32 integer rather than redrawn — the behaviour the ULID specification
/// calls monotonic. This is NOT required by the PRD; the PRD does not mention
/// it at all (see the library comment above).
///
/// Both the clock and the entropy source are injectable so tests can pin them.
class UlidGenerator {
  /// Creates a generator.
  ///
  /// [now] returns epoch-millis and defaults to the system clock. [random]
  /// defaults to [Random.secure]; a seeded [Random] makes output reproducible
  /// in tests, and must never be used in production — a predictable message id
  /// is a predictable idempotency key.
  UlidGenerator({int Function()? now, Random? random})
      : _now = now ?? _systemNowMillis,
        _random = random ?? Random.secure();

  static int _systemNowMillis() => DateTime.now().millisecondsSinceEpoch;

  final int Function() _now;
  final Random _random;

  /// Epoch-millis of the last ULID minted, or -1 before the first.
  int _lastMillis = -1;

  /// The random half of the last ULID, as 16 digits each in 0..31.
  final List<int> _lastRandom = List<int>.filled(_randomLength, 0);

  /// Mints the next ULID.
  ///
  /// Throws [StateError] in the (practically unreachable) case that the random
  /// half overflows after 32^16 calls inside a single millisecond. Throwing
  /// beats wrapping: a wrapped counter mints a duplicate id, and under D1 a
  /// duplicate id is a message the server dedups away — data loss reported as
  /// success. A crash is the honest outcome.
  String next() {
    final int millis = _now();

    if (millis == _lastMillis) {
      _incrementLastRandom();
    } else {
      // A clock that jumps BACKWARD (NTP correction, user changing the device
      // time — routine on a phone) is treated the same as any forward jump:
      // the random half is redrawn and `_lastMillis` follows the clock. The
      // ids stop being globally sorted, which is unavoidable once the clock
      // lies, but they stay unique and stay valid. Ordering on the wire is
      // `seq` (D2) and is unaffected.
      _lastMillis = millis;
      for (int i = 0; i < _randomLength; i++) {
        _lastRandom[i] = _random.nextInt(32);
      }
    }

    return _encodeTime(millis) + _encodeRandom();
  }

  /// Adds one to `_lastRandom`, read as a big-endian base-32 integer.
  void _incrementLastRandom() {
    for (int i = _randomLength - 1; i >= 0; i--) {
      if (_lastRandom[i] < 31) {
        _lastRandom[i]++;
        return;
      }
      _lastRandom[i] = 0;
    }
    throw StateError(
      'ULID random component overflowed within a single millisecond',
    );
  }

  String _encodeRandom() {
    final StringBuffer out = StringBuffer();
    for (int i = 0; i < _randomLength; i++) {
      out.write(_encoding[_lastRandom[i]]);
    }
    return out.toString();
  }
}

/// Encodes epoch-millis as 10 Crockford base32 characters, most significant
/// first (which is what makes ULIDs sort by time as plain strings).
///
/// Uses division and modulo rather than bit shifts on purpose. Dart's `>>` and
/// `&` are 32-bit when compiled to JavaScript, and a 48-bit millisecond
/// timestamp does not fit in 32 bits — a shift-based encoder is correct on
/// Android and iOS and silently wrong on Flutter Web. Division is correct on
/// all three.
String _encodeTime(int millis) {
  int remaining = millis;
  final List<String> chars = List<String>.filled(_timeLength, '0');
  for (int i = _timeLength - 1; i >= 0; i--) {
    final int mod = remaining % 32;
    chars[i] = _encoding[mod];
    remaining = (remaining - mod) ~/ 32;
  }
  return chars.join();
}
